cocotb/runner.py

```diff
@@ -73,187 +111,236 @@ def set_env(self) -> None:
-            results_xml_name = "results.xml"
+            results_xml_name = results_xml
 
-        results_xml_file = os.getenv(
-            "COCOTB_RESULTS_FILE", os.path.join(self.build_dir, results_xml_name)
-        )
-
-        self.env["COCOTB_RESULTS_FILE"] = results_xml_file
+        results_xml_file = Path(self.test_dir) / results_xml_name
 
         with suppress(OSError):
             os.remove(results_xml_file)
 
-        cmds = self.test_command()
-        self.set_env()
-        self.execute(cmds, cwd=self.sim_dir)
+        # transport the settings to cocotb via environment variables
+        self._set_env()
+        self.env["COCOTB_RESULTS_FILE"] = str(results_xml_file)
 
-        check_results_file(results_xml_file)
+        cmds: Sequence[Command] = self._test_command()
+        self._execute(cmds, cwd=self.test_dir)
+
+        # Only when running under pytest, check the results file here,
+        # potentially raising an exception with failing testcases,
+        # otherwise return the results file for later analysis.
+        if pytest_current_test:
+            check_results_file(results_xml_file)
 
         print(f"INFO: Results file: {results_xml_file}")
         return results_xml_file
 
-    @abc.abstractmethod
-    def get_include_options(self, includes: Sequence[str]) -> List[str]:
-        """Return simulator options setting directories searched for Verilog include files."""
+    @staticmethod
+    def _get_include_options(self, includes: Sequence[PathLike]) -> Command:
+        """Return simulator-specific formatted option strings with *includes* directories."""
 
         raise NotImplementedError()
 
-    @abc.abstractmethod
-    def get_define_options(self, defines: Sequence[str]) -> List[str]:
-        """Return simulator options defining macros."""
+    @staticmethod
+    def _get_define_options(self, defines: Mapping[str, object]) -> Command:
+        """Return simulator-specific formatted option strings with *defines* macros."""
 
         raise NotImplementedError()
 
     @abc.abstractmethod
-    def get_parameter_options(self, parameters: Mapping[str, object]) -> List[str]:
-        """Return simulator options for module parameters/generics."""
+    def _get_parameter_options(self, parameters: Mapping[str, object]) -> Command:
+        """Return simulator-specific formatted option strings with *parameters*/generics."""
 
         raise NotImplementedError()
 
-    def execute(self, cmds: Sequence[Command], cwd: PathLike) -> None:
+    def _execute(self, cmds: Sequence[Command], cwd: PathLike) -> None:
 
         __tracebackhide__ = True  # Hide the traceback when using PyTest.
 
         for cmd in cmds:
-            print(
-                "INFO: Running command: "
-                + ' "'.join(cmd)
-                + '" in directory:"'
-                + str(cwd)
-                + '"'
-            )
+            print(f"INFO: Running command {' '.join(cmd)} in directory {cwd}")
```

Comment:
To get the shell quoting right, define a polyfill for shlex.join(), e.g. 

```
def shlex_join(split_command):
    """Return a shell-escaped string from *split_command*."""
    return ' '.join(quote(arg) for arg in split_command)
```

(From https://github.com/python/cpython/blob/main/Lib/shlex.py)

and then
```suggestion
            print(f"INFO: Running command {shlex_join(cmd)} in directory {cwd}")
```

Reply:
One need also add.
```
_find_unsafe = re.compile(r'[^\w@%+=:,./-]', re.ASCII).search

def quote(s):
    """Return a shell-escaped version of the string *s*."""
    if not s:
        return "''"
    if _find_unsafe(s) is None:
        return s

    # use single quotes, and put single quotes into double quotes
    # the string $'b is then quoted as '$'"'"'b'
    return "'" + s.replace("'", "'\"'\"'") + "'"
```

Should I add this?

Reply:
May I ask why shlex.quote/shlex.join is not used?

Reply:
@tianrui-wei 
>May I ask why shlex.quote/shlex.join is not used?

If is supported in python 3.6 than probably no good reason. Can you propose a PR?

Reply:
@themperek PR at https://github.com/cocotb/cocotb/pull/3150